osu.Game.Rulesets.Taiko/Difficulty/Evaluators/ColourEvaluator.cs

```diff
@@ -0,0 +1,67 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using osu.Game.Rulesets.Difficulty.Preprocessing;
+using osu.Game.Rulesets.Taiko.Difficulty.Preprocessing;
+using osu.Game.Rulesets.Taiko.Difficulty.Preprocessing.Colour;
+using osu.Game.Rulesets.Taiko.Difficulty.Preprocessing.Colour.Data;
+
+namespace osu.Game.Rulesets.Taiko.Difficulty.Evaluators
+{
+    public class ColourEvaluator
+    {
+        /// <summary>
+        /// A sigmoid function. It gives a value between (middle - height/2) and (middle + height/2).
+        /// </summary>
+        /// <param name="val">The input value.</param>
+        /// <param name="center">The center of the sigmoid, where the largest gradient occurs and value is equal to middle.</param>
+        /// <param name="width">The radius of the sigmoid, outside of which values are near the minimum/maximum.</param>
+        /// <param name="middle">The middle of the sigmoid output.</param>
+        /// <param name="height">The height of the sigmoid output. This will be equal to max value - min value.</param>
+        private static double sigmoid(double val, double center, double width, double middle, double height)
+        {
+            double sigmoid = Math.Tanh(Math.E * -(val - center) / width);
+            return sigmoid * (height / 2) + middle;
+        }
+
+        /// <summary>
+        /// Evaluate the difficulty of the first note of a <see cref="MonoEncoding"/>.
+        /// </summary>
+        public static double EvaluateDifficultyOf(MonoEncoding encoding)
+        {
+            return sigmoid(encoding.Index, 2, 2, 0.5, 1) * EvaluateDifficultyOf(encoding.Parent!) * 0.5;
```

Comment:
I don't know why `encoding.Parent` is nullable and the nullability is elided here again when it is basically expected that the parent should be here after the preprocessing is done preparing the structure. Would prefer to remove these exclamation points, make `Parent` non-nullable where applicable and initialise with `null!` instead.

Reply:
This comment still hasn't been addressed, for whatever that's worth.

Reply:
Didn't know it could be declared that way, applied in fb9bb2d42dd8884cfd4d93e15024d791a69e2eda (was awaiting the answer to another comment to avoid potential conflicts when pushing)